File: backend/app/schemas/audit.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class AuditActor(BaseModel):
    user_id: int | None = None
    email: str | None = None
    role: str | None = None


class AuditClient(BaseModel):
    ip: str | None = None
    user_agent: str | None = None


class AuditResource(BaseModel):
    type: str | None = None
    id: str | None = None


class AuditLogItem(BaseModel):
    id: int
    timestamp: datetime
    event_type: str
    outcome: str
    request_id: str | None = None

    actor: AuditActor
    client: AuditClient
    resource: AuditResource

    metadata: dict[str, Any]


class AuditLogListResponse(BaseModel):
    items: list[AuditLogItem]

    total: int
    offset: int
    limit: int


class AuditLogSummaryResponse(BaseModel):
    total_events: int
    successful_events: int
    failed_events: int
    blocked_events: int
    events_last_24_hours: int

File: frontend/src/types/audit.ts
export type AuditActor = {
  user_id: number | null;
  email: string | null;
  role: string | null;
};


export type AuditClient = {
  ip: string | null;
  user_agent: string | null;
};


export type AuditResource = {
  type: string | null;
  id: string | null;
};


export type AuditLogItem = {
  id: number;
  timestamp: string;
  event_type: string;
  outcome: string;
  request_id: string | null;

  actor: AuditActor;
  client: AuditClient;
  resource: AuditResource;

  metadata: Record<
    string,
    unknown
  >;
};


export type AuditLogListResponse = {
  items: AuditLogItem[];

  total: number;
  offset: number;
  limit: number;
};


export type AuditLogSummary = {
  total_events: number;
  successful_events: number;
  failed_events: number;
  blocked_events: number;
  events_last_24_hours: number;
};


export type AuditLogFilters = {
  event_type?: string;
  outcome?: string;
  actor_email?: string;
  actor_role?: string;
  resource_type?: string;
};